[Bug]

## 🐞 버그 설명

@AuthenticationPrincipal User 방식 사용 시 NullPointerException 발생


---

## ✅ 재현 방법

1. 로그인하여 토큰 발급
2. 주문 생성 API 호출 (POST /api/users/orders)
3. 내부적으로 @AuthenticationPrincipal User를 통해 유저 객체 주입 시도
4. SecurityContext에 userId만 저장된 구조로 인해 NPE 발생

---

## 🔍 기대 동작
유저 ID 기반으로 User 객체를 명시적으로 조회하고, 정상적으로 주문이 생성되어야 함

---

## 🖥️ 환경 정보 (가능하면 작성)
- 운영체제(OS): macOS 15
- 브라우저: ex) Chrome 125.0.0.1, Safari 17  
- 디바이스: Laptop
- 실행 환경: 로컬

---

## 📸 스크린샷 (선택)
```
Caused by: java.lang.NullPointerException: Cannot invoke "String.getBytes()" because the return value of "io.groom.scubadive.shoppingmall.global.securirty.JwtProperties.getSecretKey()" is null
	at io.groom.scubadive.shoppingmall.global.securirty.JwtTokenProvider.init(JwtTokenProvider.java:22)

```

---

## 📎 참고 자료
- 해결 방안: @LoginUser Long userId로 교체 후 userService.getUserById(userId) 호출 구조로 통일

예시:

--- a/src/main/java/io/groom/scubadive/shoppingmall/global/exception/ErrorCode.java
+++ b/src/main/java/io/groom/scubadive/shoppingmall/global/exception/ErrorCode.java
@@ -19,6 +19,7 @@ public enum ErrorCode {
     NO_CHANGES_REQUESTED(HttpStatus.BAD_REQUEST, "변경된 정보가 없습니다."),
     INVALID_NICKNAME(HttpStatus.BAD_REQUEST,"유효하지 않은 닉네임입니다."),
     INVALID_PHONE_NUMBER(HttpStatus.BAD_REQUEST,"유효하지 않은 전화번호입니다."),
+    ALREADY_COMPLETED_ORDER(HttpStatus.BAD_REQUEST,"이미 완료된 주문입니다."),
 
     // 401 UNAUTHORIZED
     UNAUTHORIZED(HttpStatus.UNAUTHORIZED, "인증되지 않았습니다."),

--- a/src/main/java/io/groom/scubadive/shoppingmall/global/securirty/JwtAuthenticationFilter.java
+++ b/src/main/java/io/groom/scubadive/shoppingmall/global/securirty/JwtAuthenticationFilter.java
@@ -1,5 +1,7 @@
 package io.groom.scubadive.shoppingmall.global.securirty;
 
+import io.groom.scubadive.shoppingmall.member.domain.User;
+import io.groom.scubadive.shoppingmall.member.repository.UserRepository;
 import jakarta.servlet.FilterChain;
 import jakarta.servlet.ServletException;
 import jakarta.servlet.http.HttpServletRequest;
@@ -20,6 +22,7 @@ import java.util.List;
 public class JwtAuthenticationFilter extends OncePerRequestFilter {
 
     private final JwtTokenProvider jwtTokenProvider;
+    private final UserRepository userRepository;
 
     @Override
     protected void doFilterInternal(HttpServletRequest request,
@@ -35,6 +38,9 @@ public class JwtAuthenticationFilter extends OncePerRequestFilter {
             if (jwtTokenProvider.validateToken(token)) {
                 Long userId = jwtTokenProvider.getUserIdFromToken(token);
                 String role = jwtTokenProvider.getRoleFromToken(token);
+
+                User user = userRepository.findById(userId)
+                        .orElseThrow(() -> new IllegalStateException("인증된 사용자를 찾을 수 없습니다."));
 
                 UsernamePasswordAuthenticationToken authentication =
                         new UsernamePasswordAuthenticationToken(

--- a/src/main/java/io/groom/scubadive/shoppingmall/member/service/UserService.java
+++ b/src/main/java/io/groom/scubadive/shoppingmall/member/service/UserService.java
@@ -168,6 +168,16 @@ public class UserService {
     }
 
     /**
+     * 사용자 엔티티 조회 (비즈니스 로직용)
+     * - 주문, 결제 등에서 User 객체가 필요할 때 사용
+     * - Service 내부 도메인 로직 처리에 활용
+     */
+    public User getUserById(Long userId) {
+        return userRepository.findById(userId)
+                .orElseThrow(() -> new GlobalException(ErrorCode.USER_NOT_FOUND));
+    }
+
+    /**
      * 내 정보 수정 - 비밀번호/닉네임/전화번호 변경
      * 변경된 항목이 없으면 예외 발생
      */

--- a/src/main/java/io/groom/scubadive/shoppingmall/order/controller/OrderController.java
+++ b/src/main/java/io/groom/scubadive/shoppingmall/order/controller/OrderController.java
@@ -1,6 +1,7 @@
 package io.groom.scubadive.shoppingmall.order.controller;
 
 import io.groom.scubadive.shoppingmall.global.dto.ApiResponseDto;
+import io.groom.scubadive.shoppingmall.global.securirty.LoginUser;
 import io.groom.scubadive.shoppingmall.member.domain.User;
 import io.groom.scubadive.shoppingmall.order.dto.request.OrderCreateRequest;
 import io.groom.scubadive.shoppingmall.order.dto.response.OrderListResponse;
@@ -31,9 +32,9 @@ public class OrderController {
     })
     @PostMapping
     public ResponseEntity<ApiResponseDto<OrderResponse>> createOrder(
-            @AuthenticationPrincipal User user,
+            @LoginUser Long userId,
             @RequestBody OrderCreateRequest request) {
-        OrderResponse response = orderService.createOrder(user, request);
+        OrderResponse response = orderService.createOrder(userId, request);
         return ResponseEntity.status(201).body(ApiResponseDto.of(201, "주문 생성 성공", response));
     }
 
@@ -55,10 +56,10 @@ public class OrderController {
     })
     @GetMapping
     public ResponseEntity<ApiResponseDto<OrderListResponse>> getUserOrders(
-            @AuthenticationPrincipal User user,
+            @LoginUser Long userId,
             @RequestParam(defaultValue = "0") int page,
             @RequestParam(defaultValue = "10") int size) {
-        OrderListResponse response = orderService.getUserOrders(user, page, size);
+        OrderListResponse response = orderService.getUserOrders(userId, page, size);
         return ResponseEntity.ok(ApiResponseDto.of(200, "사용자 주문 목록 조회 성공", response));
     }
 }

--- a/src/main/java/io/groom/scubadive/shoppingmall/order/service/OrderService.java
+++ b/src/main/java/io/groom/scubadive/shoppingmall/order/service/OrderService.java
@@ -3,6 +3,8 @@ package io.groom.scubadive.shoppingmall.order.service;
 import io.groom.scubadive.shoppingmall.cart.domain.Cart;
 import io.groom.scubadive.shoppingmall.cart.domain.CartItem;
 import io.groom.scubadive.shoppingmall.cart.repository.CartRepository;
+import io.groom.scubadive.shoppingmall.global.exception.ErrorCode;
+import io.groom.scubadive.shoppingmall.global.exception.GlobalException;
 import io.groom.scubadive.shoppingmall.member.domain.User;
 import io.groom.scubadive.shoppingmall.member.service.UserPaidService;
 import io.groom.scubadive.shoppingmall.member.service.UserService;
@@ -30,8 +32,10 @@ public class OrderService {
     private final UserService userService;
 
     @Transactional
-    public OrderResponse createOrder(User user, OrderCreateRequest request) {
-        Cart cart = cartRepository.findById(request.getCartId()).orElseThrow();
+    public OrderResponse createOrder(Long userId, OrderCreateRequest request) {
+        User user = userService.getUserById(userId);
+        Cart cart = cartRepository.findById(request.getCartId())
+                .orElseThrow(() -> new IllegalArgumentException("장바구니를 찾을 수 없습니다."));
 
         if (cart.getItems().isEmpty()) throw new IllegalStateException("장바구니가 비어있습니다.");
 
@@ -58,8 +62,8 @@ public class OrderService {
         return mapToOrderResponse(order);
     }
 
-    public OrderListResponse getUserOrders(User user, int page, int size) {
-        Page<Order> orders = orderRepository.findAllByUserId(user.getId(), PageRequest.of(page, size));
+    public OrderListResponse getUserOrders(Long userId, int page, int size) {
+        Page<Order> orders = orderRepository.findAllByUserId(userId, PageRequest.of(page, size));
         return OrderListResponse.builder()
                 .page(page)
                 .totalPages(orders.getTotalPages())

--- a/src/main/java/io/groom/scubadive/shoppingmall/order/service/OrderService.java
+++ b/src/main/java/io/groom/scubadive/shoppingmall/order/service/OrderService.java
@@ -98,6 +98,9 @@ public class OrderService {
     public void changeStatus(Long orderId, OrderStatus status) {
         Order order = orderRepository.findById(orderId).orElseThrow();
         OrderStatus previousStatus = order.getStatus();
+        if (previousStatus == OrderStatus.COMPLETED) {
+            throw new GlobalException(ErrorCode.ALREADY_COMPLETED_ORDER);
+        }
 
         order.changeStatus(status);
 

